network_simulator/plot_tool.py

```diff
@@ -0,0 +1,300 @@
+
+    def graph_links(self):
+        """
+        Graphs the link stats of each available link.
+        """
+        # Set up legend storage for link names.
+        occpy_legend = []
+        loss_legend = []
+        trans_legend = []
+
+        # Iterate through each link to display each stats.
+        for link, link_stats in self.stats.link_stats.iteritems():
+            # Name of link would be "link_name".
+            # Buffer occupancy w.r.t time.
+            plt.subplot(431)
+            self.tool.graph_tuple_list(link_stats.buffer_occupancy)
+            occpy_legend.append(link)
+
+            # Packet loss times.
+            plt.subplot(434)
+            loss_rate_lst = self.tool.gen_count_interval_list(
+                link_stats.packet_loss_times)
+            self.tool.graph_tuple_list(loss_rate_lst)
+            loss_legend.append(link)
+
+            # Packet transmission w.r.t time.
+            plt.subplot(437)
+            self.tool.graph_tuple_list(link_stats.packet_transmit_times)
+            trans_legend.append(link)
+
+        # Finalize buffer occupancy graph.
+        plt.subplot(431)
+        plt.legend(occpy_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Buffer Occupancy (pkts)")
+        # Set up grid on graph.
+        plt.grid(True)
+
+        # Finalize loss times graph.
+        plt.subplot(434)
+        plt.legend(loss_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Losses")
+        # Set up grid on graph.
+        plt.grid(True)
+
+        # Finalize packet transmission graph.
+        plt.subplot(437)
+        plt.legend(trans_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Size of Packet Transmitted (Bits)")
+        # Set up grid on graph.
+        plt.grid(True)
+
+    def graph_flows(self):
+        """
+        Graphs the flow stats of each available flow.
+        """
+        # Set up legend storage for flow names.
+        sent_legend = []
+        receive_legend = []
+        rtt_legend = []
+        window_legend = []
+
+        # Iterate through each flow to display each stats.
+        for flow, flow_stats in self.stats.flow_stats.iteritems():
+            # Name of flow would be "flow_id".
+            # Packet sent times.
+            plt.subplot(432)
+            sent_rate_lst = self.tool.gen_rate_tuple_list(
+                flow_stats.packet_sent_times)
+            self.tool.graph_tuple_list(sent_rate_lst)
+            sent_legend.append(flow)
+
+            # Packet receive times.
+            plt.subplot(435)
+            rec_rate_lst = self.tool.gen_rate_tuple_list(
+                flow_stats.packet_rec_times)
+            self.tool.graph_tuple_list(rec_rate_lst)
+            receive_legend.append(flow)
+
+            # Packet RTT times.
+            plt.subplot(438)
+            self.tool.graph_1d_list(flow_stats.packet_rtts)
```

Comment:
Packet RTTs is a plot of RTT (in sec) vs time (in sec) when that RTT was received. It is not a 1D plot. You might have to change the "flow_packet_received" function and the description for "packet_rtts" so that packet_rtts is actually a list of (timestamp, RTT) tuples, where timestamp is the curr_time when the RTT was received.